v2/blockstore/ro_blockstore.go

```diff
@@ -0,0 +1,190 @@
+func ReadOnlyOf(backing io.ReaderAt, index index.Index) *ReadOnlyBlockStore {
+	return &ReadOnlyBlockStore{backing, index}
+}
+
+// LoadReadOnly opens a read-only blockstore, generating an index if it does not exist
+func LoadReadOnly(path string, noPersist bool) (*ReadOnlyBlockStore, error) {
+	reader, err := mmap.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	idx, err := index.Restore(path)
+	if err != nil {
+		idx, err = index.GenerateIndex(reader, 0, index.IndexSorted)
+		if err != nil {
+			return nil, err
+		}
+		if !noPersist {
+			if err = index.Save(idx, path); err != nil {
+				return nil, err
+			}
+		}
+	}
+	obj := ReadOnlyBlockStore{
+		backing: reader,
+		idx:     idx,
+	}
+	return &obj, nil
+}
+
+func (b *ReadOnlyBlockStore) read(idx int64) (cid.Cid, []byte, error) {
+	bcid, data, err := util.ReadNode(bufio.NewReader(internalio.NewOffsetReader(b.backing, idx)))
+	return bcid, data, err
+}
+
+// DeleteBlock is unsupported and always returns an error
+func (b *ReadOnlyBlockStore) DeleteBlock(_ cid.Cid) error {
+	return errUnsupported
+}
+
+// Has indicates if the store has a cid
+func (b *ReadOnlyBlockStore) Has(key cid.Cid) (bool, error) {
+	offset, err := b.idx.Get(key)
+	if err != nil {
+		return false, err
+	}
+	uar := internalio.NewOffsetReader(b.backing, int64(offset))
+	_, err = binary.ReadUvarint(uar)
+	if err != nil {
+		return false, err
+	}
+	c, _, err := internalio.ReadCid(b.backing, uar.Offset())
+	if err != nil {
+		return false, err
+	}
+	return c.Equals(key), nil
+}
+
+// Get gets a block from the store
+func (b *ReadOnlyBlockStore) Get(key cid.Cid) (blocks.Block, error) {
+	offset, err := b.idx.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	entry, bytes, err := b.read(int64(offset))
+	if err != nil {
+		// TODO Improve error handling; not all errors mean NotFound.
+		return nil, blockstore.ErrNotFound
+	}
+	if !entry.Equals(key) {
+		return nil, blockstore.ErrNotFound
+	}
+	return blocks.NewBlockWithCid(bytes, key)
+}
+
+// GetSize gets how big a item is
+func (b *ReadOnlyBlockStore) GetSize(key cid.Cid) (int, error) {
+	idx, err := b.idx.Get(key)
+	if err != nil {
+		return -1, err
+	}
+	l, err := binary.ReadUvarint(internalio.NewOffsetReader(b.backing, int64(idx)))
+	if err != nil {
+		return -1, blockstore.ErrNotFound
+	}
+	c, _, err := internalio.ReadCid(b.backing, int64(idx+l))
+	if err != nil {
+		return 0, err
+	}
+	if !c.Equals(key) {
+		return -1, blockstore.ErrNotFound
+	}
+	// get cid. validate.
+	return int(l), err
+}
+
+// Put is not supported and always returns an error
+func (b *ReadOnlyBlockStore) Put(blocks.Block) error {
+	return errUnsupported
+}
+
+// PutMany is not supported and always returns an error
+func (b *ReadOnlyBlockStore) PutMany([]blocks.Block) error {
+	return errUnsupported
+}
+
+// AllKeysChan returns the list of keys in the store
+func (b *ReadOnlyBlockStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
```

Comment:
we may use this walk for populating the index, and we need to be able to iterate keys in this way somewhere for index generation. In general though, when it's asked for all keys from a blockstore with an index, we should iterate through the index when possible rather than linear reads through the full car.

Reply:
Added TODO thank you for pointing this out